pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +126,156 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if !cr.Spec.SmartUpdateEnabled() {
+		return nil
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
+	if err != nil {
+		return fmt.Errorf("failed to get sate: %v", err)
+	}
+
+	if currentSet.Generation == startGeneration {
+		return nil
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	return r.smartUpdate(sfs, cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api.PerconaXtraDBCluster) error {
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	list := corev1.PodList{}
+	err = r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	)
+	if err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	var primaryPod *corev1.Pod = nil
+	for _, pod := range list.Items {
+		pod := pod
```

Comment:
just Q: what the sense of this line?

Reply:
Because we later get reference of pod it can cause a lot of tricky bugs. This line "pins" `pod` var
https://github.com/kyoh86/scopelint
https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
https://www.tddfellow.com/blog/2016/01/08/why-do-you-need-to-be-careful-with-loop-variable-in-go/#taking_reference_of_loop_variable